File: bin/sp-4.4.1/sicstus-4.4.1/se/sics/prologbeans/PBListCell.java
/**
 * Copyright (c) 2003 SICS AB. All rights reserved.
 */
package se.sics.prologbeans;

import java.io.IOException;

/**
 * <code>PBListCell</code> is the 
 * representation of Prolog list cells, i.e. compound terms with the functor
 * ./2.
 */
class PBListCell extends PBCompound {
	// [PM] FIXME: synchronization story
	private String cachedString = null;
	// [PM] FIXME: synchronization story
	private Boolean cachedIsTrueString = null;

	PBListCell(PBTerm head, PBTerm tail) {
		this(new PBTerm[] { head, tail });
	}

	PBListCell(PBTerm[] terms) {
		super(".", terms);
		// assert terms != null && terms.length == 2;
	}

	public boolean isListCell() {
		return true;
	}

	// There is no Character.isDefined(int) in .NET
	private boolean isUnicode(int c) {
		return ((c < 0xD800 || c > 0xDFFF) && c <= 0x10FFFD);
	}

	// There is no StringBuilder.appendCodePoint(int) in .NET
	private void appendCodePoint(StringBuffer sb, int c) {
		if (c < 0xFFFF) {
			sb.append((char) c);
		} else {
			int c1 = c - 0x10000;
			int c1h = c1 >> 10; // higher 10 bits of c1
			int c1l = (c1 << 10) >> 10; // lower 10 bits of c1
			int u1 = 0xD800;
			int u2 = 0xDC00;
			u1 = u1 | c1h;
			u2 = u2 | c1l;
			sb.append((char) u1);
			sb.append((char) u2);
		}
	}

	public boolean isString() {
		if (cachedString != null) {
			return true;
		} else {
			try {
				cachedString = getString();
				return true;
			} catch (IllegalStateException ise) {
				cachedString = null; // *** FIXME: is this necessary?
				return false;
			}
		}
	}

	/**
	 * Returns <code>true</code> if this <code>PBTerm</code> is a proper list
	 * and all of its elements are character codes. Returns <code>false</code>
	 * otherwise.
	 */
	private boolean isTrueString() {
		if (cachedIsTrueString == null) {
			boolean trueString = true;
			PBListCell pblc;
			PBTerm h, t = this;
			while ((t != PBTerm.NIL) && trueString) {
				pblc = (PBListCell) t;
				h = pblc.arguments[0]; // head()
				if (h.isInteger() && isUnicode((int) h.intValue())) {
					t = pblc.arguments[1]; // tail()
					trueString = (t.isListCell() || t == PBTerm.NIL);
				} else {
					trueString = false;
				}
			}
			cachedIsTrueString = new Boolean(trueString);
		}
		return cachedIsTrueString.booleanValue();
	}

	public String getString() // throws IllegalStateException
	{
		if (cachedString == null) {
			StringBuffer sbuf = new StringBuffer();
			cachedString = getString(sbuf).toString();
		}
		return cachedString;
	}

	// [PD] 4.0.0 Non-recursive version
	private StringBuffer getString(StringBuffer sbuf) // throws
														// IllegalStateException
	{
		PBListCell pbt = this;
		PBTerm h;
		do {
			h = pbt.arguments[0];
			if (h.isAtom() && h.name.length() == 1) { // Is head a one char
														// atom?
				sbuf.append(((PBAtom) h).getName());
			} else if (h.isInteger()) { // Is head an integer?
				// *** FIXME: Perhaps check if this cast truncates.
				int c = (int) h.intValue();
				if (!isUnicode(c)) {
					throw new IllegalStateException("not a list of characters");
				}
				appendCodePoint(sbuf, c);
			} else {
				throw new IllegalStateException("not a list of characters");
			}
			PBTerm t = pbt.arguments[1]; // tail()
			if (t == PBTerm.NIL) {
				return sbuf;
			} else {
				if (t.isListCell()) {
					pbt = (PBListCell) t;
				} else {
					throw new IllegalStateException("not a list of characters");
				}
			}
		} while (true);
	}

	/**
	 * Returns the head of this list cell.
	 */
	public PBTerm head() {
		// return getArgument(1);
		return arguments[0];
	}

	/**
	 * Returns the tail of this list cell.
	 */
	public PBTerm tail() {
		// return getArgument(2);
		return arguments[1];
	}

	/**
	 * Returns the length of this <code>PBListCell</code>.
	 * <p>
	 */
	public int length() {
		// [PM] 4.1.3 non-recursive version
		int len = 0;
		PBTerm tmp;
		for (tmp = this; tmp.isListCell(); tmp = tmp.tail()) {
			len++;
		}
		// [PM] 4.1.3 for verifying properness. Expect tmp to be a PBNil
		// here, i.e. tmp.length()==0.
		len += tmp.length();
		// return 1 + tail().length();
		return len;
	}

	String toPrologString() {
		StringBuffer sb = new StringBuffer().append('[');
		sb.append(arguments[0].toPrologString());
		PBTerm t = arguments[1];
		while (!t.isEmptyList()) {
			sb.append(',');
			sb.append(t.head().toPrologString());
			t = t.tail();
		}
		sb.append(']');
		return sb.toString();
	}

	public String toString() {
		StringBuffer sb = new StringBuffer().append('[');
		sb.append(arguments[0].toString());
		PBTerm t = arguments[1];
		while (!t.isEmptyList()) {
			sb.append(',');
			sb.append(t.head().toString());
			t = t.tail();
		}
		sb.append(']');
		return sb.toString();
	}

	void fastWrite(FastWriter writer) throws IOException {
		if (isTrueString()) {
			writer.writeString(getString());
		} else {
			super.fastWrite(writer);
		}
	}

	// @Override
	PBTerm[] fastWritePrefix(FastWriter writer) throws IOException {
		if (isTrueString()) {
			writer.writeString(getString());
			return PBTerm.NO_TERMS;
		} else {
			return super.fastWritePrefix(writer);
		}
	}

}

File: bin/sp-4.4.1/sicstus-4.4.1/library/prologbeans.NET/PBListCell.cs
using System;

namespace se.sics.prologbeans
{
	
	/// <summary> <code>PBListCell</code> is the 
	/// representation of Prolog list cells, i.e. compound terms with the functor
	/// ./2.
	/// </summary>
	class PBListCell:PBCompound
	{
		override public sealed bool ListCell {
			get {
				return true;
			}
		}

		override public bool ProperList {
			get {
				if (cachedIsProperList == CACHED_BOOLEAN_UNKNOWN) {
					PBTerm tail = this;
					do {
						tail = tail.getArgument (2);
					} while (tail.ListCell);

					bool isProperList = tail.EmptyList;

					cachedIsProperList = (isProperList ? CACHED_BOOLEAN_TRUE : CACHED_BOOLEAN_FALSE);
				}
				return cachedIsProperList == CACHED_BOOLEAN_TRUE;
			}	
		}

		/// <summary> Returns <code>true</code> if this <code>PBTerm</code> is a proper list
		/// and all of its elements are character codes. Returns <code>false</code>
		/// otherwise.
		/// </summary>
		bool TrueString {
			get {
				// [PM] 4.3.2 Changed to use three byte values {unknown, false, true}.
				if (cachedIsTrueString == CACHED_BOOLEAN_UNKNOWN) {
					bool trueString = true;
					PBListCell pblc;
					PBTerm h, t = this;
					while ((t != PBTerm.NIL) && trueString) {
						pblc = (PBListCell)t;
						h = pblc.arguments [0]; // head()
						if (h.Integer && isUnicode ((int)h.intValue ())) {
							t = pblc.arguments [1]; // tail()
							trueString = (t.ListCell || t == PBTerm.NIL);
						} else {
							trueString = false;
						}
					}
					cachedIsTrueString = (trueString ? CACHED_BOOLEAN_TRUE : CACHED_BOOLEAN_FALSE);
				}
				return cachedIsTrueString == CACHED_BOOLEAN_TRUE;
			}
		}

		// [PM] FIXME: synchronization story
		string cachedString = null;
		// [PM] FIXME: synchronization story

		const byte CACHED_BOOLEAN_UNKNOWN = 0;
		const byte CACHED_BOOLEAN_FALSE = 1;
		const byte CACHED_BOOLEAN_TRUE = 2;
		byte cachedIsTrueString = CACHED_BOOLEAN_UNKNOWN;
		byte cachedIsProperList = CACHED_BOOLEAN_UNKNOWN;

		internal PBListCell (PBTerm head, PBTerm tail) : this (new []{ head, tail })
		{
		}

		internal PBListCell (PBTerm[] terms) : base (".", terms)
		{
			// assert terms != null && terms.length == 2;
		}
		
		// There is no Character.isDefined(int) in .NET
		static bool isUnicode (int c)
		{
			return ((c < 0xD800 || c > 0xDFFF) && c <= 0x10FFFD);
		}
		
		// There is no StringBuilder.appendCodePoint(int) in .NET
		static void appendCodePoint (System.Text.StringBuilder sb, int c)
		{
			if (c < 0xFFFF) {
				sb.Append ((char)c);
			} else {
				int c1 = c - 0x10000;
				int c1h = c1 >> 10; // higher 10 bits of c1
				int c1l = (c1 << 10) >> 10; // lower 10 bits of c1
				int u1 = 0xD800;
				int u2 = 0xDC00;
				u1 = u1 | c1h;
				u2 = u2 | c1l;
				sb.Append ((char)u1);
				sb.Append ((char)u2);
			}
		}

		public override bool isString ()
		{
			if (cachedString != null) {
				return true;
			} else {
				try {
					cachedString = getString ();
					return true;
				} catch (SystemException /* ise */) {
					cachedString = null; // *** FIXME: is this necessary?
					return false;
				}
			}
		}

		public override string getString ()
		{
			if (cachedString == null) {
				System.Text.StringBuilder sbuf = new System.Text.StringBuilder ();
				cachedString = getString (sbuf).ToString ();
			}
			return cachedString;
		}
		
		// [PD] 4.0.0 Non-recursive version
		System.Text.StringBuilder getString (System.Text.StringBuilder sbuf)
		{
			PBListCell pbt = this;
			PBTerm h;
			do {
				h = pbt.arguments [0];
				if (h.Atom && h.Name.Length == 1) {
					// Is head a one char
					// atom?
					sbuf.Append (h.Name);
				} else if (h.Integer) {
					// Is head an integer?
					// *** FIXME: Perhaps check if this cast truncates.
					int c = (int)h.intValue ();
					if (!isUnicode (c)) {
						throw new InvalidOperationException ("not a list of characters");
					}
					appendCodePoint (sbuf, c);
				} else {
					throw new InvalidOperationException ("not a list of characters");
				}
				PBTerm t = pbt.arguments [1]; // tail()
				if (t == PBTerm.NIL) {
					return sbuf;
				} else {
					if (t.ListCell) {
						pbt = (PBListCell)t;
					} else {
						throw new InvalidOperationException ("not a list of characters");
					}
				}
			} while (true);
		}

		/// <summary> Returns the head of this list cell.</summary>
		public override PBTerm head ()
		{
			return arguments [0];
		}

		/// <summary> Returns the tail of this list cell.</summary>
		public override PBTerm tail ()
		{
			return arguments [1];
		}

		public override int length ()
		{
			// [PM] 4.1.3 non-recursive version
			int len = 0;
			PBTerm tmp;
			for (tmp = this; tmp.ListCell; tmp = tmp.tail ()) {
				len++;
			}
			// [PM] 4.1.3 for verifying properness. Expect tmp to be a PBNil
			// here, i.e. tmp.length()==0.
			len += tmp.length ();
			return len;
		}

		internal override string toPrologString ()
		{
			System.Text.StringBuilder sb = new System.Text.StringBuilder ().Append ('[');
			sb.Append (arguments [0].toPrologString ());
			PBTerm t = arguments [1];
			while (!t.EmptyList) {
				sb.Append (',');
				sb.Append (t.head ().toPrologString ());
				t = t.tail ();
			}
			sb.Append (']');
			return sb.ToString ();
		}

		public override string ToString ()
		{
			System.Text.StringBuilder sb = new System.Text.StringBuilder ().Append ('[');
			sb.Append (arguments [0].ToString ());
			PBTerm t = arguments [1];
			while (!t.EmptyList) {
				sb.Append (',');
				sb.Append (t.head ().ToString ());
				t = t.tail ();
			}
			sb.Append (']');
			return sb.ToString ();
		}

		internal override PBTerm[] fastWritePrefix (FastWriter writer)
		{
			// [PM] 4.3.2 FIXME: it would be better to always write a string prefix (writeCharList) if possible, and then write the remainder regardless of whether the remainder is NIL or not.
			if (TrueString) {
				writer.writeString (getString ());
				return PBTerm.NO_TERMS;
			} else {
				return base.fastWritePrefix (writer);
			}
		}
	}
}